driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java

```diff
@@ -104,29 +103,37 @@ Session session() {
         return manager.getSession();
     }
 
-    PreparedStatement getPreparedQuery(QueryType type, Set<ColumnMapper<?>> columns, EnumMap<Option.Type, Option> options) {
-
-        MapperQueryKey pqk = new MapperQueryKey(type, columns, options);
+    ListenableFuture<PreparedStatement> getPreparedQueryAsync(QueryType type, Set<ColumnMapper<?>> columns, EnumMap<Option.Type, Option> options) {
 
+        final MapperQueryKey pqk = new MapperQueryKey(type, columns, options);
         PreparedStatement stmt = preparedQueries.get(pqk);
         if (stmt == null) {
-            synchronized (preparedQueries) {
-                stmt = preparedQueries.get(pqk);
-                if (stmt == null) {
-                    String queryString = type.makePreparedQueryString(tableMetadata, mapper, manager, columns, options.values());
-                    logger.debug("Preparing query {}", queryString);
-                    stmt = session().prepare(queryString);
-                    Map<MapperQueryKey, PreparedStatement> newQueries = new HashMap<MapperQueryKey, PreparedStatement>(preparedQueries);
-                    newQueries.put(pqk, stmt);
-                    preparedQueries = newQueries;
+            String queryString = type.makePreparedQueryString(tableMetadata, mapper, manager, columns, options.values());
+            logger.debug("Preparing query {}", queryString);
+            final SettableFuture<PreparedStatement> future = SettableFuture.create();
+            Futures.addCallback(session().prepareAsync(queryString), new FutureCallback<PreparedStatement>() {
```

Comment:
:+1: , created [JAVA-1096](https://datastax-oss.atlassian.net/browse/JAVA-1096) for this and referenced this discussion thread.